File: back/src/main/java/com/back/global/security/oauth2/service/CustomOAuth2UserService.java
package com.back.global.security.oauth2.service;

import com.back.domain.user.entity.LoginType;
import com.back.domain.user.entity.User;
import com.back.domain.user.repository.UserRepository;
import com.back.global.security.oauth2.info.GoogleOAuth2UserInfo;
import com.back.global.security.oauth2.info.KakaoOAuth2UserInfo;
import com.back.global.security.oauth2.info.OAuth2UserInfo;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.oauth2.core.user.DefaultOAuth2User;
import org.springframework.security.oauth2.core.user.OAuth2User;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Map;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
@Transactional
public class CustomOAuth2UserService extends DefaultOAuth2UserService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    @Override
    public OAuth2User loadUser(OAuth2UserRequest userRequest) {
        OAuth2User oAuth2User = super.loadUser(userRequest);

        String registrationId = userRequest
                .getClientRegistration()
                .getRegistrationId();

        String refreshToken = userRequest.getAdditionalParameters()
                .getOrDefault("refresh_token", "").toString();

        log.info("OAuth2 additionalParameters: {}", userRequest.getAdditionalParameters());
        log.info("OAuth2 refreshToken: {}", refreshToken);

        User user = switch (registrationId) {
            case "kakao" -> getOrCreateUser(oAuth2User.getAttributes(), LoginType.KAKAO, refreshToken);
            case "google" -> getOrCreateUser(oAuth2User.getAttributes(), LoginType.GOOGLE, refreshToken);
            default -> throw new OAuth2AuthenticationException("oauth2_provider_not_supported");
        };

        return new DefaultOAuth2User(
                oAuth2User.getAuthorities(),
                Map.of(
                        "userId", user.getUserId(),
                        "loginId", user.getLoginId(),
                        "email", user.getEmail(),
                        "name", user.getName()
                ),
                "userId"
        );
    }

    private OAuth2UserInfo createUserInfo(
            Map<String, Object> attributes,
            LoginType loginType
    ) {
        return switch (loginType) {
            case KAKAO -> new KakaoOAuth2UserInfo(attributes);
            case GOOGLE -> new GoogleOAuth2UserInfo(attributes);
            default -> throw new OAuth2AuthenticationException("oauth2_provider_not_supported");
        };
    }

    private User getOrCreateUser(Map<String, Object> attributes, LoginType loginType, String refreshToken) {
        OAuth2UserInfo userInfo = createUserInfo(attributes, loginType);

        String platformId = userInfo.getProviderId();
        String loginId = loginType.name() + "_" + platformId;
        String email = userInfo.getEmail();
        String name = userInfo.getName();

        validateRequired(platformId, "oauth2_provider_id_missing");
        validateRequired(email, "oauth2_email_missing");

        return userRepository.findByLoginIdAndDeletedAtIsNull(loginId)
                .map(user -> {
                    if (!refreshToken.isBlank()) {
                        user.updateOauthRefreshToken(refreshToken);
                    }
                    return user;
                })
                .orElseGet(() -> createOAuthUser(loginId, email, name, loginType, refreshToken));
    }

    private User createOAuthUser(
            String loginId,
            String email,
            String name,
            LoginType loginType,
            String refreshToken
    ) {
        if (userRepository.existsByEmailAndDeletedAtIsNull(email)) {
            throw new OAuth2AuthenticationException("oauth2_email_already_exists");
        }

        String randomPassword = passwordEncoder.encode(UUID.randomUUID().toString());

        User user = User.createOAuth(
                loginId,
                email,
                randomPassword,
                name,
                loginType,
                refreshToken.isBlank() ? null : refreshToken
        );

        try {
            return userRepository.save(user);
        } catch (DataIntegrityViolationException e) {
            throw new OAuth2AuthenticationException("oauth2_email_already_exists");
        }
    }

    private void validateRequired(String value, String errorCode) {
        if (value == null || value.isBlank()) {
            throw new OAuth2AuthenticationException(errorCode);
        }
    }
}

File: back/src/main/kotlin/com/back/global/security/oauth2/service/CustomOAuth2UserService.kt
package com.back.global.security.oauth2.service

import com.back.domain.user.constant.LoginType
import com.back.domain.user.entity.User
import com.back.domain.user.repository.UserRepository
import com.back.domain.auth.service.SocialLinkService
import com.back.domain.auth.repository.SocialLinkCookieRepository
import com.back.global.security.oauth2.info.GoogleOAuth2UserInfo
import com.back.global.security.oauth2.info.KakaoOAuth2UserInfo
import com.back.global.security.oauth2.info.NaverOAuth2UserInfo
import com.back.global.security.oauth2.info.OAuth2UserInfo
import org.slf4j.LoggerFactory
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest
import org.springframework.security.oauth2.core.OAuth2AuthenticationException
import org.springframework.security.oauth2.core.user.DefaultOAuth2User
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
@Transactional
class CustomOAuth2UserService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val socialLinkService: SocialLinkService,
    private val socialLinkCookieRepository: SocialLinkCookieRepository,
) : DefaultOAuth2UserService() {

    override fun loadUser(userRequest: OAuth2UserRequest): OAuth2User {
        val oAuth2User = super.loadUser(userRequest)
        val registrationId = userRequest.clientRegistration.registrationId
        val refreshToken = userRequest.additionalParameters.getOrDefault("refresh_token", "").toString()

        @Suppress("UNCHECKED_CAST")
        val attributes = oAuth2User.attributes as Map<String, Any>

        val (loginType, userInfo) = when (registrationId) {
            "kakao" -> LoginType.KAKAO to createUserInfo(attributes, LoginType.KAKAO)
            "naver" -> LoginType.NAVER to createUserInfo(oAuth2User.attributes, LoginType.NAVER)
            "google" -> LoginType.GOOGLE to createUserInfo(attributes, LoginType.GOOGLE)
            else -> throw OAuth2AuthenticationException("oauth2_provider_not_supported")
        }
        val linkIntentId = socialLinkCookieRepository.load()
        val isSocialLink = linkIntentId != null
        val user = linkIntentId
            ?.let { intentId ->
                socialLinkService.complete(
                    intentId = intentId,
                    provider = loginType,
                    userInfo = userInfo,
                    oauthRefreshToken = refreshToken.ifBlank { null },
                )
            }
            ?: getOrCreateUser(userInfo, loginType, refreshToken)

        val userId = user.userId ?: throw IllegalStateException("User ID must not be null")

        return DefaultOAuth2User(
            oAuth2User.authorities,
            mapOf(
                "userId" to userId,
                "loginId" to user.loginId,
                "email" to user.email,
                "name" to user.name,
                "socialLink" to isSocialLink,
            ),
            "userId"
        )
    }

    private fun createUserInfo(attributes: Map<String, Any>, loginType: LoginType): OAuth2UserInfo =
        when (loginType.name) {
            "KAKAO" -> KakaoOAuth2UserInfo(attributes)
            "NAVER" -> NaverOAuth2UserInfo(attributes)
            "GOOGLE" -> GoogleOAuth2UserInfo(attributes)
            else -> throw OAuth2AuthenticationException("oauth2_provider_not_supported")
        }

    private fun getOrCreateUser(userInfo: OAuth2UserInfo, loginType: LoginType, refreshToken: String): User {
        val platformId = userInfo.providerId
        val email = userInfo.email
        val name = userInfo.name

        validateRequired(platformId, "oauth2_provider_id_missing")
        validateRequired(email, "oauth2_email_missing")

        val safePlatformId = platformId ?: throw OAuth2AuthenticationException("oauth2_provider_id_missing")
        val safeEmail = email ?: throw OAuth2AuthenticationException("oauth2_email_missing")
        val loginId = "${loginType.name}_$safePlatformId"

        val existingUser =
            userRepository.findBySocialProviderAndSocialProviderIdAndDeletedAtIsNull(
                loginType,
                safePlatformId,
            )
                ?: userRepository.findByLoginIdAndDeletedAtIsNull(loginId)
        if (existingUser != null) {
            if (existingUser.socialProvider == null) {
                existingUser.linkSocialAccount(
                    loginType,
                    safePlatformId,
                    refreshToken.ifBlank { null },
                )
            }
            if (refreshToken.isNotBlank()) {
                existingUser.updateOauthRefreshToken(refreshToken)
            }
            return existingUser
        }

        return createOAuthUser(loginId, safeEmail, name, loginType, safePlatformId, refreshToken)
    }

    private fun createOAuthUser(
        loginId: String,
        email: String,
        name: String,
        loginType: LoginType,
        providerId: String,
        refreshToken: String
    ): User {
        if (userRepository.existsByEmailAndDeletedAtIsNull(email)) {
            throw OAuth2AuthenticationException("oauth2_email_already_exists")
        }

        val encodedPassword = passwordEncoder.encode(UUID.randomUUID().toString())
        val randomPassword = requireNotNull(encodedPassword) { "Password encoding failed" }

        val user = User.createOAuth(
            loginId = loginId,
            email = email,
            password = randomPassword,
            name = name,
            loginType = loginType,
            providerId = providerId,
            oauthRefreshToken = refreshToken.ifBlank { null }
        )

        return try {
            userRepository.save(user)
        } catch (e: DataIntegrityViolationException) {
            throw OAuth2AuthenticationException("oauth2_email_already_exists")
        }
    }

    private fun validateRequired(value: String?, errorCode: String) {
        if (value.isNullOrBlank()) {
            throw OAuth2AuthenticationException(errorCode)
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(CustomOAuth2UserService::class.java)
    }
}
